supplementary/alica_ros_turtlesim/ros2/src/turtle_ros2_interfaces.cpp

```diff
@@ -0,0 +1,122 @@
+#include "alica_ros2_turtlesim/turtle_ros2_interfaces.hpp"
+
+#include <engine/blackboard/Blackboard.h>
+
+#include <geometry_msgs/msg/twist.hpp>
+#include <rclcpp/logging.hpp>
+#include <turtlesim/srv/spawn.hpp>
+#include <turtlesim/srv/teleport_absolute.hpp>
+
+#include <chrono>
+#include <future>
+using namespace std::chrono_literals;
+
+namespace turtlesim
+{
+
+TurtleRos2Interfaces::TurtleRos2Interfaces(const std::string& name)
+        : TurtleInterfaces(name)
+        , _name(name)
+{
+    // initialize publisher, subscriber and service client.
+    _initTrigger = false;
+
+    _nh = rclcpp::Node::make_shared("alica_ros2_turtlesim");
+    _spinner.add_node(_nh);
+    _nh->declare_parameter("name", "Hello!");
+
+    rclcpp::CallbackGroup::SharedPtr teleportCallback = _nh->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
+    _nh->get_parameter("name", _name);
+
+    // initialize publisher, subscriber and service client.
+    _velPub = _nh->create_publisher<geometry_msgs::msg::Twist>("/" + _name + "/cmd_vel", 10);
+    _initTriggerSub = _nh->create_subscription<std_msgs::msg::Empty>("/init", 10, [&](const std_msgs::msg::Empty& msg) { _initTrigger = true; });
+    _poseSub = _nh->create_subscription<turtlesim::msg::Pose>("/" + _name + "/pose", 10, [&](const msg::Pose::ConstSharedPtr msg) { _current = msg; });
+    _teleportClient = _nh->create_client<turtlesim::srv::TeleportAbsolute>("/" + _name + "/teleport_absolute", rmw_qos_profile_default, teleportCallback);
+
+    rclcpp::CallbackGroup::SharedPtr spawnCallback = _nh->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
+    _spawnClient = _nh->create_client<turtlesim::srv::Spawn>("/spawn", rmw_qos_profile_services_default, spawnCallback);
+    _spinThread = std::thread([this]() { _spinner.spin(); });
```

Comment:
not clear for me how this thread should stop execution and why it's not joined. (same above but OOS)

Reply:
We usually kill the app with ctrl+c and dont do a proper clean up, maybe we should cancel the spinner and join the thread when we receive the kill signal.

Reply:
I guess ROS has a ctrl+c handler which not immediately kills the whole system and trying to shut it down properly
It could be great if we could make it more aligned with C++ standard